### PR TITLE
[Vulkan] Add initial support of AOT

### PR DESCRIPTION
**Pull request by ghuau-innopeak**
Related issue = https://github.com/taichi-dev/taichi/issues/3642

### CODE REVIEW

**Comment by ghuau-innopeak** [comment]
There is one issue with the current PR I'm not entirely sure: why it's needed to run `lower(kernel);` in `AotModuleBuilderImpl::add_per_backend(...)`, otherwise, we don't get an OffloadStatement but a FrontendStatement

**Comment by k-ye** [comment]
> There is one issue with the current PR I'm not entirely sure: why it's needed to run `lower(kernel);` in `AotModuleBuilderImpl::add_per_backend(...)`, otherwise, we don't get an OffloadStatement but a FrontendStatement

Yup, `lower()` is the mechanism to translate the frontend AST to the lower-level, SSA compliant IR (namely CHI IR)

File: taichi/backends/vulkan/aot_module_builder_impl.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+class AotModuleBuilderImpl : public AotModuleBuilder {
+ public:
+  explicit AotModuleBuilderImpl(
+      VkRuntime *runtime,
+      const std::vector<CompiledSNodeStructs> &compiled_structs);
+  void dump(const std::string &output_dir,
+            const std::string &filename) const override;
+
+ protected:
```
Comment (k-ye) [inline taichi/backends/vulkan/aot_module_builder_impl.h:25]: nit: I belive this class is not gonna be further inherited.
```suggestion
 private:
```

**Review by k-ye — COMMENTED** [review]
Thanks! The PR looks good, I'll leave this to @ailzhang for additional comments, if any :-)

File: taichi/backends/vulkan/aot_module_builder_impl.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    : compiled_structs_(compiled_structs), runtime_(runtime) {
+}
+
+void AotModuleBuilderImpl::write_spv_file(
+    const std::string &output_dir,
+    const std::string &filename,
+    const TaskAttributes &k,
+    const std::vector<uint32_t> &source_code) const {
+  const std::string spv_path =
+      fmt::format("{}/{}_{}.spv", output_dir, filename, k.name);
```
Comment (ailzhang) [inline taichi/backends/vulkan/aot_module_builder_impl.cpp:24]: btw we recently found out that ignoring the filename prefix is probably more convenient (https://github.com/taichi-dev/taichi/pull/3648), I'm curious if you want the same for vulkan backend ;)

**Review by ailzhang — APPROVED** [review]
Looks great! Thanks!!

File: taichi/backends/vulkan/aot_module_builder_impl.h (hunk 1 above)
Comment (ghuau-innopeak) [reply taichi/backends/vulkan/aot_module_builder_impl.h]: I agree, I moved it to private and if it needs to be inherited and used later... we can still change it back to protected :).

File: taichi/backends/vulkan/aot_module_builder_impl.cpp (hunk 2 above)
Comment (ghuau-innopeak) [reply taichi/backends/vulkan/aot_module_builder_impl.cpp]: Definitely! I pushed the same changes to this PR, that would avoid to do multiple one :). I tested it and seems to work as well.

**Review by k-ye — APPROVED** [review]

**Pull request merged**